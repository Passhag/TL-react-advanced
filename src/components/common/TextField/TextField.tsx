import * as React from 'react';

interface TextFieldProps {
  onFocus?: Function;
  onBlur?: Function;
  autoFocus?: boolean;
  autoComplete?: 'on' | 'off';
  placeholder?: string;
  name: string;
  label: string;
}

interface TextFieldState {
  isFocused?: boolean;
}

export default class TextFieldComponent
  extends React.Component<TextFieldProps, TextFieldState> {
  public static defaultProps: Partial<TextFieldProps> = {
    autoFocus: true,
    autoComplete: 'off',
    placeholder: '',
  };

  constructor(props: TextFieldProps) {
    super(props);

    this.state = {
      isFocused: false,
    };

    this._onFocus = this._onFocus.bind(this);
    this._onBlur = this._onBlur.bind(this);
  }

  public render(): JSX.Element {
    const {
      label,
      name,
      autoFocus,
      autoComplete,
      placeholder,
      onBlur,
      onFocus,
    } = this.props;

    return (
      <div className={`tl-text-field-wrapper ${this.state.isFocused && 'active'}`}>
        <label className="tl-text-field-label" htmlFor={name}>
          {label}
        </label>
        <input
          className="tl-text-field-input"
          type="text"
          name={name}
          autoFocus={autoFocus}
          autoComplete={autoComplete}
          placeholder={placeholder}
          onFocus={() => { onFocus ? (onFocus(), this._onFocus()) : this._onFocus(); }}
          onBlur={() => { onBlur ? (onBlur(), this._onBlur()) : this._onBlur(); }}
        />
        <span className="tl-text-field-separator" />
      </div>
    );
  }

  private _onFocus(): void {
    this.setState({
      isFocused: true,
    });
  }

  private _onBlur(): void {
    this.setState({
      isFocused: false,
    });
  }
}
