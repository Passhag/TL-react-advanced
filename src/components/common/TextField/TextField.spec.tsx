import * as React from 'react';
import { shallow, configure, ShallowWrapper } from 'enzyme';
import * as ReactSixteenAdapter from 'enzyme-adapter-react-16';
import TextField from './TextField';

configure({ adapter: new ReactSixteenAdapter() });

const aShallowWrapper = (props?: { [key: string]: Function }): ShallowWrapper => (
  shallow(<TextField name="test" {...props} />)
);

describe('Components::TextField', () => {
  it('should render correctly', () => {
    const wrapper = aShallowWrapper();

    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger onKeyDown listener', () => {
    const onKeyDown = jasmine.createSpy('onKeyDown');
    const wrapper = aShallowWrapper({ onKeyDown });

    wrapper.find('input').simulate('keydown');

    expect(onKeyDown).toHaveBeenCalled();
  });

  it('should trigger onChange listener', () => {
    const onChange = jasmine.createSpy('onChange');
    const wrapper = aShallowWrapper({ onChange });

    wrapper.find('input').simulate('change');

    expect(onChange).toHaveBeenCalled();
  });

  it('should trigger onFocus listener', () => {
    const onFocus = jasmine.createSpy('onFocus');
    const wrapper = aShallowWrapper({ onFocus });

    wrapper.find('input').simulate('focus');

    expect(onFocus).toHaveBeenCalled();
  });

  it('should trigger onBlur listener', () => {
    const onBlur = jasmine.createSpy('onBlur');
    const wrapper = aShallowWrapper({ onBlur });

    wrapper.find('input').simulate('blur');

    expect(onBlur).toHaveBeenCalled();
  });
});
