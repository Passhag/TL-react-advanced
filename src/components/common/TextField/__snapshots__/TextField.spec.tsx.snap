// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::TextField should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextFieldComponent
    autoComplete="off"
    autoFocus={true}
    name="test"
    placeholder=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "autoComplete": "off",
      "autoFocus": true,
      "className": "form-control",
      "name": "test",
      "placeholder": "",
      "type": "text",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "autoComplete": "off",
        "autoFocus": true,
        "className": "form-control",
        "name": "test",
        "placeholder": "",
        "type": "text",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
